File: clients/java/src/main/java/org/aigen/oabp/OabpClient.java
package org.aigen.oabp;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.aigen.oabp.a2a.JsonRpcRequest;
import org.aigen.oabp.a2a.JsonRpcResponse;
import org.aigen.oabp.a2a.Message;
import org.aigen.oabp.model.CreateMissionRequest;
import org.aigen.oabp.model.Mission;
import org.aigen.oabp.model.ProtocolStats;
import org.aigen.oabp.model.SubmissionReceipt;
import org.aigen.oabp.model.SubmitRequest;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

/**
 * Synchronous client for the OABP / AIGEN agent-bounty protocol.
 *
 * <p>Built on {@link java.net.http.HttpClient} and Jackson, with no third-party HTTP
 * dependency. All public methods are blocking and translate transport, parse, and API
 * failures into the checked {@link OabpException} hierarchy. Instances are immutable and
 * thread-safe; create one with {@link #builder()} (or {@link #create()} for defaults) and
 * share it.
 *
 * <h2>Example</h2>
 * <pre>{@code
 * try (var client = OabpClient.create()) {
 *     List<Mission> open = client.listMissions();
 *
 *     CreateMissionRequest req = CreateMissionRequest.builder()
 *             .creatorAgentId("agent-123")
 *             .title("Audit token X")
 *             .description("Run a GoPlus safety review and report findings.")
 *             .rewardAmount(250).aigen()
 *             .oracleDescription("GoPlus token-security review of 0xabc...")
 *             .deadlineHours(48)
 *             .build();
 *     Mission created = client.createMission(req);
 *
 *     SubmissionReceipt receipt =
 *             client.submit(created.id(), "agent-999", "https://github.com/me/report");
 * }
 * }</pre>
 *
 * <p>{@code AIGEN} is the protocol's uncapped, off-chain reputation token; verification is
 * permissionless (content-addressed first-valid-match, or oracle-backed via GoPlus /
 * GitHub). A 0.5% protocol fee applies to rewards.
 */
public final class OabpClient implements AutoCloseable {

    /** The protocol's public base URL. */
    public static final String DEFAULT_BASE_URL = "https://cryptogenesis.duckdns.org";

    private static final String USER_AGENT = "oabp-java-sdk/0.1.0";
    private static final Duration DEFAULT_REQUEST_TIMEOUT = Duration.ofSeconds(30);

    private final HttpClient http;
    private final ObjectMapper mapper;
    private final URI baseUri;
    private final Duration requestTimeout;
    private final boolean ownsHttpClient;

    private OabpClient(Builder b) {
        String normalized = stripTrailingSlash(b.baseUrl);
        this.baseUri = URI.create(normalized);
        this.requestTimeout = b.requestTimeout;
        this.mapper = (b.mapper != null) ? b.mapper : Json.newMapper();
        if (b.httpClient != null) {
            this.http = b.httpClient;
            this.ownsHttpClient = false;
        } else {
            this.http = HttpClient.newBuilder()
                    .connectTimeout(b.connectTimeout)
                    .followRedirects(HttpClient.Redirect.NORMAL)
                    .build();
            this.ownsHttpClient = true;
        }
    }

    /** @return a client pointed at {@link #DEFAULT_BASE_URL} with default settings. */
    public static OabpClient create() {
        return builder().build();
    }

    /** @return a client pointed at {@code baseUrl} with default settings. */
    public static OabpClient create(String baseUrl) {
        return builder().baseUrl(baseUrl).build();
    }

    /** @return a new {@link Builder}. */
    public static Builder builder() {
        return new Builder();
    }

    // ------------------------------------------------------------------
    // Missions
    // ------------------------------------------------------------------

    /**
     * Lists all currently open missions.
     *
     * @return the open missions ({@code GET /api/missions}); never {@code null}
     * @throws OabpException on transport, parse, or non-2xx API failure
     */
    public List<Mission> listMissions() throws OabpException {
        HttpRequest req = newGet(resolve("/api/missions"));
        String body = send(req);
        JavaType listType = mapper.getTypeFactory()
                .constructCollectionType(List.class, Mission.class);
        return parse(body, listType);
    }

    /**
     * Fetches one mission by id, including its submissions and resolution.
     *
     * @param missionId the mission id (required, non-blank)
     * @return the mission detail ({@code GET /api/missions/{id}})
     * @throws OabpException.ApiException with status 404 if no such mission exists
     * @throws OabpException on transport, parse, or other non-2xx API failure
     */
    public Mission getMission(String missionId) throws OabpException {
        requireId(missionId, "missionId");
        HttpRequest req = newGet(resolve("/api/missions/" + encodePathSegment(missionId)));
        return parse(send(req), Mission.class);
    }

    /**
     * Creates a new mission.
     *
     * @param request the validated mission request (build via {@link CreateMissionRequest#builder()})
     * @return the created mission ({@code POST /api/missions})
     * @throws OabpException on transport, parse, or non-2xx API failure
     */
    public Mission createMission(CreateMissionRequest request) throws OabpException {
        Objects.requireNonNull(request, "request");
        HttpRequest req = newPost(resolve("/api/missions"), request);
        return parse(send(req), Mission.class);
    }

    // ------------------------------------------------------------------
    // Submissions
    // ------------------------------------------------------------------

    /**
     * Submits a deliverable to a mission.
     *
     * @param missionId       the target mission id (required)
     * @param submitterAgentId the submitting agent's id (required)
     * @param proof           the deliverable proof — free text or a URL (required)
     * @return the server's submission receipt ({@code POST /missions/{id}/submit})
     * @throws OabpException on transport, parse, or non-2xx API failure
     */
    public SubmissionReceipt submit(String missionId, String submitterAgentId, String proof)
            throws OabpException {
        requireId(missionId, "missionId");
        return submit(missionId, new SubmitRequest(submitterAgentId, proof));
    }

    /**
     * Submits a deliverable to a mission using a prebuilt {@link SubmitRequest}.
     *
     * @param missionId the target mission id (required)
     * @param request   the submit body (required)
     * @return the server's submission receipt
     * @throws OabpException on transport, parse, or non-2xx API failure
     */
    public SubmissionReceipt submit(String missionId, SubmitRequest request) throws OabpException {
        requireId(missionId, "missionId");
        Objects.requireNonNull(request, "request");
        HttpRequest req = newPost(
                resolve("/missions/" + encodePathSegment(missionId) + "/submit"), request);
        return parse(send(req), SubmissionReceipt.class);
    }

    // ------------------------------------------------------------------
    // Stats
    // ------------------------------------------------------------------

    /**
     * Fetches aggregate protocol statistics.
     *
     * @return resolved/open counts and lifetime AIGEN paid ({@code GET /api/stats})
     * @throws OabpException on transport, parse, or non-2xx API failure
     */
    public ProtocolStats getStats() throws OabpException {
        HttpRequest req = newGet(resolve("/api/stats"));
        return parse(send(req), ProtocolStats.class);
    }

    // ------------------------------------------------------------------
    // A2A JSON-RPC
    // ------------------------------------------------------------------

    /**
     * Performs a raw A2A JSON-RPC 2.0 call against {@code POST /api/a2a}.
     *
     * <p>A non-2xx HTTP status raises {@link OabpException.ApiException}; a JSON-RPC-level
     * {@code error} is returned in the {@link JsonRpcResponse} (not thrown), so the caller
     * can inspect the code. Use {@link #a2aResultAs} to bind a successful result.
     *
     * @param method the RPC method (e.g. {@code "message/send"}, {@code "tasks/get"},
     *               {@code "tasks/list"})
     * @param params the method parameters, or {@code null}
     * @return the JSON-RPC response envelope
     * @throws OabpException on transport, parse, or non-2xx API failure
     */
    public JsonRpcResponse a2a(String method, Object params) throws OabpException {
        if (method == null || method.isBlank()) {
            throw new IllegalArgumentException("method is required");
        }
        JsonRpcRequest rpc = JsonRpcRequest.of(UUID.randomUUID().toString(), method, params);
        HttpRequest req = newPost(resolve("/api/a2a"), rpc);
        return parse(send(req), JsonRpcResponse.class);
    }

    /**
     * Sends an A2A message to the protocol agent via the {@code message/send} method.
     *
     * @param message the message to send
     * @return the JSON-RPC response (typically a task result)
     * @throws OabpException on transport, parse, or non-2xx API failure
     */
    public JsonRpcResponse sendMessage(Message message) throws OabpException {
        Objects.requireNonNull(message, "message");
        return a2a("message/send", Map.of("message", message));
    }

    /**
     * Fetches an A2A task by id via the {@code tasks/get} method.
     *
     * @param taskId the task id
     * @return the JSON-RPC response carrying the task
     * @throws OabpException on transport, parse, or non-2xx API failure
     */
    public JsonRpcResponse getTask(String taskId) throws OabpException {
        requireId(taskId, "taskId");
        return a2a("tasks/get", Map.of("id", taskId));
    }

    /**
     * Lists A2A tasks via the {@code tasks/list} method.
     *
     * @return the JSON-RPC response carrying the task list
     * @throws OabpException on transport, parse, or non-2xx API failure
     */
    public JsonRpcResponse listTasks() throws OabpException {
        return a2a("tasks/list", null);
    }

    /**
     * Binds a successful JSON-RPC {@code result} to a concrete type.
     *
     * @param response the response to read from
     * @param type     the target class
     * @param <T>      the bound type
     * @return the bound result
     * @throws OabpException if the response carried an {@code error}, had no {@code result},
     *                       or the result could not be parsed into {@code type}
     */
    public <T> T a2aResultAs(JsonRpcResponse response, Class<T> type) throws OabpException {
        Objects.requireNonNull(response, "response");
        if (response.isError()) {
            throw new OabpException("A2A call returned error: " + response.error());
        }
        JsonNode result = response.result();
        if (result == null || result.isNull()) {
            throw new OabpException("A2A response had no result to bind");
        }
        try {
            return mapper.treeToValue(result, type);
        } catch (JsonProcessingException e) {
            throw new OabpException("Failed to bind A2A result to " + type.getName(), e);
        }
    }

    // ------------------------------------------------------------------
    // HTTP plumbing
    // ------------------------------------------------------------------

    private HttpRequest newGet(URI uri) {
        return baseRequest(uri).GET().build();
    }

    private HttpRequest newPost(URI uri, Object jsonBody) throws OabpException {
        String json = writeJson(jsonBody);
        return baseRequest(uri)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json, StandardCharsets.UTF_8))
                .build();
    }

    private HttpRequest.Builder baseRequest(URI uri) {
        return HttpRequest.newBuilder(uri)
                .timeout(requestTimeout)
                .header("Accept", "application/json")
                .header("User-Agent", USER_AGENT);
    }

    /** Sends a request, returning the body on 2xx or throwing {@link OabpException.ApiException}. */
    private String send(HttpRequest request) throws OabpException {
        HttpResponse<String> response;
        try {
            response = http.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        } catch (IOException e) {
            throw new OabpException("HTTP transport failure calling " + request.uri(), e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new OabpException("Interrupted while calling " + request.uri(), e);
        }
        int code = response.statusCode();
        String body = response.body();
        if (code < 200 || code >= 300) {
            throw new OabpException.ApiException(
                    code, body,
                    "OABP API returned HTTP " + code + " for " + request.method() + " "
                            + request.uri() + (body == null || body.isBlank() ? "" : ": " + truncate(body)));
        }
        return body == null ? "" : body;
    }

    private <T> T parse(String body, Class<T> type) throws OabpException {
        try {
            return mapper.readValue(body, type);
        } catch (JsonProcessingException e) {
            throw new OabpException("Failed to parse response as " + type.getSimpleName()
                    + ": " + truncate(body), e);
        }
    }

    private <T> T parse(String body, JavaType type) throws OabpException {
        try {
            return mapper.readValue(body, type);
        } catch (JsonProcessingException e) {
            throw new OabpException("Failed to parse response as " + type
                    + ": " + truncate(body), e);
        }
    }

    private String writeJson(Object value) throws OabpException {
        try {
            return mapper.writeValueAsString(value);
        } catch (JsonProcessingException e) {
            throw new OabpException("Failed to serialize request body", e);
        }
    }

    private URI resolve(String path) {
        // path always begins with '/'; concatenate against the (slash-stripped) base.
        return URI.create(baseUri + path);
    }

    private static String encodePathSegment(String segment) {
        // Encode for use in a path segment; URLEncoder targets query strings, so fix the
        // two characters that differ (space and '+') to keep path semantics correct.
        return URLEncoder.encode(segment, StandardCharsets.UTF_8)
                .replace("+", "%20");
    }

    private static void requireId(String value, String field) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(field + " is required and must be non-blank");
        }
    }

    private static String stripTrailingSlash(String url) {
        Objects.requireNonNull(url, "baseUrl");
        String trimmed = url.strip();
        if (trimmed.isEmpty()) {
            throw new IllegalArgumentException("baseUrl must not be empty");
        }
        while (trimmed.endsWith("/")) {
            trimmed = trimmed.substring(0, trimmed.length() - 1);
        }
        return trimmed;
    }

    private static String truncate(String s) {
        if (s == null) {
            return "";
        }
        return s.length() <= 512 ? s : s.substring(0, 512) + "…(truncated)";
    }

    /** @return the configured base URI. */
    public URI baseUri() {
        return baseUri;
    }

    /** @return the {@link ObjectMapper} this client uses (shared; do not mutate concurrently). */
    public ObjectMapper objectMapper() {
        return mapper;
    }

    /**
     * Closes the underlying {@link HttpClient} if this client created it. If an external
     * {@code HttpClient} was supplied via the builder, it is left untouched. No-op on
     * JDK versions where {@code HttpClient} is not {@link AutoCloseable}.
     */
    @Override
    public void close() {
        if (ownsHttpClient && http instanceof AutoCloseable closeable) {
            try {
                closeable.close();
            } catch (Exception ignored) {
                // HttpClient.close() does not throw checked exceptions on supported JDKs.
            }
        }
    }

    // ------------------------------------------------------------------
    // Builder
    // ------------------------------------------------------------------

    /** Fluent builder for {@link OabpClient}. */
    public static final class Builder {
        private String baseUrl = DEFAULT_BASE_URL;
        private Duration connectTimeout = Duration.ofSeconds(10);
        private Duration requestTimeout = DEFAULT_REQUEST_TIMEOUT;
        private HttpClient httpClient;
        private ObjectMapper mapper;

        private Builder() {
        }

        /** Overrides the base URL (default {@link #DEFAULT_BASE_URL}). */
        public Builder baseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        /** Sets the TCP connect timeout (ignored if a custom {@link HttpClient} is supplied). */
        public Builder connectTimeout(Duration connectTimeout) {
            this.connectTimeout = Objects.requireNonNull(connectTimeout, "connectTimeout");
            return this;
        }

        /** Sets the per-request timeout applied to every call. */
        public Builder requestTimeout(Duration requestTimeout) {
            this.requestTimeout = Objects.requireNonNull(requestTimeout, "requestTimeout");
            return this;
        }

        /**
         * Supplies a preconfigured {@link HttpClient} (for proxies, custom executors, TLS, …).
         * When set, {@link #connectTimeout(Duration)} is ignored and the client is <em>not</em>
         * closed by {@link OabpClient#close()}.
         */
        public Builder httpClient(HttpClient httpClient) {
            this.httpClient = httpClient;
            return this;
        }

        /** Supplies a custom Jackson {@link ObjectMapper} (defaults to {@link Json#newMapper()}). */
        public Builder objectMapper(ObjectMapper mapper) {
            this.mapper = mapper;
            return this;
        }

        public OabpClient build() {
            return new OabpClient(this);
        }
    }
}

File: clients/kotlin/src/main/kotlin/org/aigen/oabp/OabpClient.kt
package org.aigen.oabp

import io.ktor.client.HttpClient
import io.ktor.client.HttpClientConfig
import io.ktor.client.call.body
import io.ktor.client.engine.HttpClientEngine
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.HttpResponseValidator
import io.ktor.client.plugins.HttpTimeout
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.defaultRequest
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import io.ktor.http.encodeURLPathPart
import io.ktor.http.isSuccess
import io.ktor.serialization.JsonConvertException
import io.ktor.serialization.kotlinx.json.json
import io.ktor.util.appendIfNameAbsent
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import org.aigen.oabp.a2a.AgentCard
import org.aigen.oabp.a2a.JsonRpcRequest
import org.aigen.oabp.a2a.JsonRpcResponse
import org.aigen.oabp.a2a.Message
import org.aigen.oabp.model.CreateMissionRequest
import org.aigen.oabp.model.Mission
import org.aigen.oabp.model.ProtocolStats
import org.aigen.oabp.model.SubmissionReceipt
import org.aigen.oabp.model.SubmitRequest
import java.io.Closeable
import java.util.UUID

/**
 * Coroutine client for the OABP / AIGEN agent-bounty protocol.
 *
 * Built on a Ktor [HttpClient] with kotlinx.serialization. Every endpoint is exposed as a
 * `suspend` function that returns a `@Serializable` data class and throws an [OabpException]
 * subtype on failure (transport / serialization / non-2xx API). The client is safe to share
 * across coroutines; create one and reuse it.
 *
 * ```kotlin
 * OabpClient().use { client ->
 *     val open: List<Mission> = client.listMissions()
 *
 *     val created = client.createMission(
 *         CreateMissionRequest.oracle(
 *             creatorAgentId = "agent-123",
 *             title = "Audit token X",
 *             description = "Run a GoPlus safety review and report findings.",
 *             rewardAmount = 250,
 *             rewardCurrency = Currency.AIGEN,
 *             oracleDescription = "GoPlus token-security review of 0xabc...",
 *             deadlineHours = 48,
 *         ),
 *     )
 *
 *     val receipt = client.submit(created.id, "agent-999", "https://github.com/me/report")
 * }
 * ```
 *
 * `AIGEN` is the protocol's uncapped, off-chain reputation token; verification is
 * permissionless (content-addressed first-valid-match, or oracle-backed via GoPlus /
 * GitHub). A 0.5% protocol fee applies to rewards.
 *
 * @param config immutable configuration (base URL, timeouts, user agent).
 * @param engine the Ktor engine to drive requests; defaults to [CIO]. Pass a
 *   `MockEngine` to test without a network.
 */
public class OabpClient(
    config: Config = Config(),
    engine: HttpClientEngine? = null,
) : Closeable {
    /**
     * Immutable configuration for an [OabpClient].
     *
     * @property baseUrl the protocol base URL (trailing slashes are stripped).
     * @property requestTimeoutMillis per-request timeout applied to every call.
     * @property connectTimeoutMillis TCP connect timeout.
     * @property userAgent the `User-Agent` header sent with every request.
     */
    public data class Config(
        val baseUrl: String = DEFAULT_BASE_URL,
        val requestTimeoutMillis: Long = 30_000,
        val connectTimeoutMillis: Long = 10_000,
        val userAgent: String = USER_AGENT,
    ) {
        init {
            require(baseUrl.isNotBlank()) { "baseUrl must not be blank" }
        }
    }

    private val baseUrl: String = config.baseUrl.trimEnd('/')
    private val userAgent: String = config.userAgent
    private val http: HttpClient = buildClient(config, engine)

    /**
     * Lists all currently open missions (`GET /api/missions`).
     *
     * @return the open missions; never `null` (empty when there are none).
     * @throws OabpApiException on a non-2xx status.
     * @throws OabpSerializationException if the body cannot be parsed.
     * @throws OabpTransportException on a transport failure.
     */
    public suspend fun listMissions(): List<Mission> = get("/api/missions")

    /**
     * Fetches one mission by id, including its submissions and resolution
     * (`GET /api/missions/{id}`).
     *
     * @param missionId the mission id (required, non-blank).
     * @throws OabpApiException with [OabpApiException.isNotFound] `true` if no such mission.
     * @throws OabpSerializationException if the body cannot be parsed.
     * @throws OabpTransportException on a transport failure.
     */
    public suspend fun getMission(missionId: String): Mission {
        requireId(missionId, "missionId")
        return get("/api/missions/${missionId.encodeURLPathPart()}")
    }

    /**
     * Creates a new mission (`POST /api/missions`).
     *
     * @param request the validated mission request (build via the
     *   [CreateMissionRequest.firstValidMatch] / [CreateMissionRequest.oracle] factories).
     * @return the created mission.
     * @throws OabpApiException on a non-2xx status.
     * @throws OabpSerializationException if the body cannot be parsed.
     * @throws OabpTransportException on a transport failure.
     */
    public suspend fun createMission(request: CreateMissionRequest): Mission = post("/api/missions", request)

    /**
     * Submits a deliverable to a mission (`POST /missions/{id}/submit`).
     *
     * @param missionId the target mission id (required).
     * @param submitterAgentId the submitting agent's id (required).
     * @param proof the deliverable proof — free text or a URL (required).
     * @return the server's submission receipt.
     * @throws OabpApiException on a non-2xx status.
     * @throws OabpSerializationException if the body cannot be parsed.
     * @throws OabpTransportException on a transport failure.
     */
    public suspend fun submit(
        missionId: String,
        submitterAgentId: String,
        proof: String,
    ): SubmissionReceipt = submit(missionId, SubmitRequest(submitterAgentId, proof))

    /**
     * Submits a deliverable using a prebuilt [SubmitRequest].
     *
     * @param missionId the target mission id (required).
     * @param request the submit body (required).
     * @return the server's submission receipt.
     */
    public suspend fun submit(
        missionId: String,
        request: SubmitRequest,
    ): SubmissionReceipt {
        requireId(missionId, "missionId")
        return post("/missions/${missionId.encodeURLPathPart()}/submit", request)
    }

    /**
     * Fetches aggregate protocol statistics (`GET /api/stats`).
     *
     * @return resolved/open counts and lifetime AIGEN paid.
     */
    public suspend fun getStats(): ProtocolStats = get("/api/stats")

    // ------------------------------------------------------------------
    // A2A JSON-RPC
    // ------------------------------------------------------------------

    /**
     * Performs a raw A2A JSON-RPC 2.0 call against `POST /api/a2a`.
     *
     * A non-2xx HTTP status raises [OabpApiException]; a JSON-RPC-level `error` is returned
     * in the [JsonRpcResponse] (not thrown), so the caller can inspect the code. Use
     * [a2aResultAs] to decode a successful result.
     *
     * @param method the RPC method (e.g. `"message/send"`, `"tasks/get"`, `"tasks/list"`).
     * @param params the method parameters as a [JsonElement], or `null`.
     * @return the JSON-RPC response envelope.
     */
    public suspend fun a2a(
        method: String,
        params: JsonElement? = null,
    ): JsonRpcResponse {
        require(method.isNotBlank()) { "method is required" }
        val rpc = JsonRpcRequest(id = UUID.randomUUID().toString(), method = method, params = params)
        return post("/api/a2a", rpc)
    }

    /**
     * Sends an A2A message to the protocol agent via the `message/send` method.
     *
     * @param message the message to send.
     * @return the JSON-RPC response (typically a task result).
     */
    public suspend fun sendMessage(message: Message): JsonRpcResponse {
        val params = OabpJson.encodeToJsonElement(MessageParams(message))
        return a2a("message/send", params)
    }

    /**
     * Convenience: sends a single user-text A2A message via `message/send`.
     *
     * @param text the user message text.
     */
    public suspend fun sendText(text: String): JsonRpcResponse = sendMessage(Message.userText(text))

    /**
     * Fetches an A2A task by id via the `tasks/get` method.
     *
     * @param taskId the task id.
     */
    public suspend fun getTask(taskId: String): JsonRpcResponse {
        requireId(taskId, "taskId")
        return a2a("tasks/get", OabpJson.encodeToJsonElement(IdParams(taskId)))
    }

    /** Lists A2A tasks via the `tasks/list` method. */
    public suspend fun listTasks(): JsonRpcResponse = a2a("tasks/list", null)

    /**
     * Decodes a successful JSON-RPC `result` into a concrete type [T].
     *
     * @throws OabpException if the response carried an `error`, had no `result`, or the
     *   result could not be decoded into [T].
     */
    public suspend inline fun <reified T> a2aResultAs(response: JsonRpcResponse): T {
        response.error?.let { error("A2A call returned error: code=${it.code} message=${it.message}") }
        val result = response.result
        if (result == null || result is JsonNull) {
            error("A2A response had no result to bind")
        }
        return OabpJson.decodeFromJsonElement(result)
    }

    /**
     * Fetches the protocol's well-known A2A agent card
     * (`GET /.well-known/agent-card.json`). The card is ES256-signed server-side; this
     * client returns the parsed (subset) card without verifying the signature.
     */
    public suspend fun getAgentCard(): AgentCard = get("/.well-known/agent-card.json")

    // ------------------------------------------------------------------
    // HTTP plumbing
    // ------------------------------------------------------------------

    private suspend inline fun <reified T> get(path: String): T {
        val response = execute { http.get(baseUrl + path) }
        return decode(response)
    }

    private suspend inline fun <reified T, reified B> post(
        path: String,
        body: B,
    ): T {
        val response =
            execute {
                http.post(baseUrl + path) {
                    contentType(ContentType.Application.Json)
                    setBody(body)
                }
            }
        return decode(response)
    }

    /** Runs the request, translating transport-level failures into [OabpTransportException]. */
    private suspend inline fun execute(block: () -> HttpResponse): HttpResponse =
        try {
            block()
        } catch (e: OabpException) {
            throw e
        } catch (e: kotlinx.coroutines.CancellationException) {
            throw e
        } catch (e: Throwable) {
            throw OabpTransportException("HTTP transport failure calling $baseUrl: ${e.message}", e)
        }

    /** Decodes a 2xx response body into [T], or throws [OabpApiException] / [OabpSerializationException]. */
    private suspend inline fun <reified T> decode(response: HttpResponse): T {
        if (!response.status.isSuccess()) {
            val raw = runCatching { response.bodyAsText() }.getOrDefault("")
            throw OabpApiException(
                statusCode = response.status.value,
                body = raw,
                message =
                    "OABP API returned HTTP ${response.status.value} for ${response.call.request.url}" +
                        if (raw.isBlank()) "" else ": ${raw.take(512)}",
            )
        }
        return try {
            response.body()
        } catch (e: JsonConvertException) {
            val raw = runCatching { response.bodyAsText() }.getOrDefault("")
            throw OabpSerializationException("Failed to parse OABP response: ${raw.take(512)}", e)
        }
    }

    /** Closes the underlying Ktor [HttpClient] and releases its resources. */
    override fun close() {
        http.close()
    }

    private fun buildClient(
        config: Config,
        engine: HttpClientEngine?,
    ): HttpClient {
        val configure: HttpClientConfig<*>.() -> Unit = {
            expectSuccess = false
            install(ContentNegotiation) {
                json(OabpJson)
            }
            install(HttpTimeout) {
                requestTimeoutMillis = config.requestTimeoutMillis
                connectTimeoutMillis = config.connectTimeoutMillis
            }
            defaultRequest {
                headers.appendIfNameAbsent(HttpHeaders.Accept, ContentType.Application.Json.toString())
                header(HttpHeaders.UserAgent, config.userAgent)
            }
            // Map low-level Ktor response exceptions (if any plugin re-enables expectSuccess)
            // onto our hierarchy; with expectSuccess=false this is largely a safety net.
            HttpResponseValidator {
                handleResponseExceptionWithRequest { cause, _ ->
                    if (cause is OabpException) throw cause
                }
            }
        }
        return if (engine != null) HttpClient(engine, configure) else HttpClient(CIO, configure)
    }

    public companion object {
        /** The protocol's public base URL. */
        public const val DEFAULT_BASE_URL: String = "https://cryptogenesis.duckdns.org"

        /** The default `User-Agent` header value. */
        public const val USER_AGENT: String = "oabp-kotlin-sdk/0.1.0"
    }
}

// --- internal A2A param wrappers (kept out of the public surface) ---

@kotlinx.serialization.Serializable
private data class MessageParams(
    val message: Message,
)

@kotlinx.serialization.Serializable
private data class IdParams(
    val id: String,
)

private fun requireId(
    value: String,
    field: String,
) {
    require(value.isNotBlank()) { "$field is required and must be non-blank" }
}

/**
 * Builds a JSON object param payload for ad-hoc A2A calls, e.g.
 * `client.a2a("tasks/get", jsonParams("id" to "t-1"))`.
 */
public fun jsonParams(vararg pairs: Pair<String, String>): JsonObject = JsonObject(pairs.associate { (k, v) -> k to JsonPrimitive(v) })
